Update FLMD and DD scripts (v2)

bpowers24 opened the pull request:
These scripts create the FLMD and DD files that fit the new ESS-DIVE standards. Additionally, this PR includes the `load_tabular_data()` function, which is primarily used before running data checks, but wasn't included in that PR because I needed to update it after I finished the flmd and dd scripts. 

New scripts that need a full review: 
- [x] `create_flmd.R` - this houses 2 functions `get_files()` and `create_flmd()`
- [x] `create_dd.R` - this houses 1 function `create_dd()`
- [x] `create_flmd_dd.R` - this is the main script you'll use to create flmd and dds
- [x] `load_tabular_data.R` - this function loads in all the tabular data

Supporting scripts that I also created: 
- [x] `example_data_for_flmd_dd_tests.R` - this is a series of functions that creates testing data in a temporary folder. Probably don't need to review this
- [x] `test-create_dd.R` - this is the test script for the DD. I would at least read through the comments so you understand the scope/constraints of the function, but don't need to run it if you don't want to. If you do run it, everything should pass. 
- [x] `test-create_flmd.R` - this is the test script for the FLMD. I would at least read through the comments so you understand the scope/constraints of the function, but don't need to run it if you don't want to. If you do run it, everything should pass. 

A couple other notes: 
- I deleted the old scripts. If you would like me to keep them and put them in an archive instead, let me know. 
- The `load_tabular_data()` function is a little clunky because of the Boye file issue with the hashtags. Once you update the Boye format, this could probably be simplified significantly. 
- Can you also look at issue #4 and see if these changes would satisfy/close that issue?

Remaining tasks for Bibi (to do after PR is approved, but before I merge it)
- [x] Update status comment in all scripts (to hopefully say "approved" lol)
- [ ] Update function name in data_checks scripts from `load_tabular_data_from_flmd()` to `load_tabular_data()`
- [ ] Run data checks to confirm the changes I made to `load_tabular_data()` didn't break anything
- [x] sort out conflicts (might have to do with the .gitignore?)

brieanne-forbes commented:
Add term type back into the dd

brieanne-forbes commented on Data_Transformation/functions/load_tabular_data.R:
```diff
@@ -1,272 +1,383 @@
+      }
+      
+      # join flmd to dir df
+      current_df_metadata <- left_join(current_df_metadata, current_flmd_df, by = c("File_Name", "File_Path_Absolute")) %>% 
+        mutate(Header_Position = coalesce(Header_Position.x, Header_Position.y)) %>% 
+        select(File_Name, Header_Position, Data_Start_Row, File_Path_Absolute)
```
current_df_metadata doesn't exist yet

brieanne-forbes commented on Data_Transformation/functions/load_tabular_data.R:
```diff
@@ -1,272 +1,383 @@
+      current_df_metadata_file_path_absolute <- tabular_metadata$File_Path_Absolute[k]
       
-      file_paths_all <- file_paths_all[file_paths_all %in% file.path(current_directory, include_files)]
+      log_info(paste0("Reading in tabular file ", k, " of ", nrow(current_tabular_only_metadata), ": ", basename(current_df_metadata_file_path_absolute)))
```
current_tabular_only_metadata doesnt exist yet

brieanne-forbes commented on Data_Transformation/functions/load_tabular_data.R:
```diff
@@ -1,272 +1,383 @@
     
-    user_input <- readline("Are all column headers on the first row? (Y/N) ")
+    # getting files that don't have a header_position
+    current_df_metadata_missing_header_position <- current_df_metadata %>% 
```
current_df_metadata doesnt exist yet

brieanne-forbes commented on Data_Transformation/functions/load_tabular_data.R:
ran into multiple errors. Will have to rerun again on next review to see if things are working properly

brieanne-forbes reviewed (changes requested):
Lots of errors and confusion this time around :/ 
General comment: All of the questions being slightly different is hard on the brain. Consider using some formatting to emphasize e.g. "What line has the DATA?"

bpowers24 commented on Data_Package_Documentation/create_flmd_dd.R:
```diff
@@ -0,0 +1,115 @@
+# exclude_files = vector of files (relative file path + file name) to exclude from within the dir. Optional argument; default is NA_character_. 
+user_exclude_files = NA_character_
+
+# include_files = vector of files (relative file path + file name) to include from within the dir. Optional argument; default is NA_character_. 
```
I have examples written in the get_files() function. Is that enough? Otherwise not sure how to write it here without making it super messy? 
https://github.com/river-corridors-sfa/rcsfa-data_processing_for_publication/blob/a4ff24210cd635fa0129d3804e4681ee88faeec1/Data_Package_Documentation/functions/create_flmd.R#L33-L48

I could add a comment that's like? 
```
# after sourcing in the function, you can run `get_files` in the console to see examples
```

bpowers24 replied:
I didn't have a test script for the get_files() function, so I didn't write a specific test to check the NA vs NA_character_ thing. But I tried running the function with both of those and it worked. 

lmk if you'd want me to create a test-get_files.R script/if that would be helpful for you in the future for debugging

bpowers24 replied:
add an example for when you'd want to use include + exclude

bpowers24 commented on Data_Package_Documentation/functions/create_dd.R:
```diff
@@ -0,0 +1,339 @@
+  
+  log_info("Extracting headers from files.")
+  
+  tabular_files_with_flmd %>% 
+    group_by(Column_or_Row_Name_Position) %>%
+    summarise(files = str_c(file, collapse = ", "), .groups = "drop") %>% 
+    print()
+  
```
I added this to indicate which rows contain the headers in each file
![image](https://github.com/user-attachments/assets/62e8deaf-a263-4883-9378-6cf43f6e9d06)

bpowers24 reviewed (commented):
Replied to some of Brie's messages

bpowers24 commented on Data_Package_Documentation/functions/create_flmd.R:
```diff
@@ -0,0 +1,450 @@
+    
+  }
+  
+  log_info(paste0("Adding ", length(current_file_paths), " of ", length(file_paths_all), " files."))
```
How about "Excluding 0 of 131 total file(s) in the directory." since I think what we care about more is knowing/confirming if any files are being removed?

bpowers24 commented on Data_Package_Documentation/functions/create_flmd.R:
```diff
@@ -0,0 +1,450 @@
+  # remove excluded files
+  if (any(!is.na(exclude_files))) {
+    
+    current_file_paths <- file_paths_all[!file_paths_all %in% file.path(directory, exclude_files)]
```
No, I double checked the code and it should be looking for exclude files within file_paths_all; lines 69 and 78 are correct. However I did refactor this code into an if/else statement to clarify that include_files() takes priority (i.e., if include_files is used, then exclude_files is skipped entirely).

bpowers24 commented on Data_Package_Documentation/functions/create_flmd.R:
```diff
@@ -0,0 +1,450 @@
+        View(current_tabular_file)
+        
+        # ask what type of header format
+        user_reply <- as.numeric(readline(prompt = cat("What type of header info is present? 0 = none; 1 = Boye; 2 = Goldman; 3 = other")))
```
can we have it print the file name and relative path and then ask the user for n, b, g, o, u = unknown. 
Use crayon package to change color of filename. Maybe purple!!
If it's none, header_rows = 1, column_or_row_name_position = 1
If it's boye, then grab the B2 cell for header rows, column_or_row_name_position = 1
If it's goldan, header_rows = 1, column_or_row_name_position = 1
If it's unknown, then show top 100 rows of file and ask for header_row, column_or_row_name_position, and standard

bpowers24 commented on Data_Transformation/functions/load_tabular_data.R:
```diff
@@ -1,272 +1,383 @@
+  # Inputs: 
+    # files_df = df with at least these 5 cols: all, absolute_dir, parent_dir, relative_dir, and file. Required argument. 
+    # flmd_df = df with at least these 3 cols: File_Name, Column_or_Row_Name_Position, File_Path. Optional argument; default is NA. 
+    # query_header_info = T/F where the user should select T if header rows are present and F if all tabular files do NOT have header rows. Select F if on NERSC. Optional argument; default is FALSE.  
```
add more documentation to explain the value of this. add a note that this should still = T even if you have added an FLMD

bpowers24 commented on Data_Package_Documentation/functions/create_flmd.R:
```diff
@@ -0,0 +1,450 @@
+      
+          # add to flmd
+          current_flmd_skeleton$Header_Rows[current_flmd_skeleton$all == current_file_absolute] <- current_header_row
+          current_flmd_skeleton$Column_or_Row_Name_Position[current_flmd_skeleton$all == current_file_absolute] <- current_column_or_row_name_position
```
599e568c29f1732cc14104731324a93c2e301a5b

bpowers24 commented on Data_Package_Documentation/functions/create_flmd.R:
```diff
@@ -0,0 +1,450 @@
+            pull(2)
+          
+          # add to flmd
+          current_flmd_skeleton$Header_Rows[current_flmd_skeleton$all == current_file_absolute] <- boye_row_info
```
599e568c29f1732cc14104731324a93c2e301a5b

bpowers24 commented on Data_Package_Documentation/functions/create_dd.R:
```diff
@@ -0,0 +1,339 @@
+      pull(Column_or_Row_Name)
+    
+    # add cols to main df
+    tabular_files$Column_or_Row_Name[tabular_files$files_flmd_join == current_files_flmd_join] <- list(current_Column_or_Row_Names)
```
leaving this one in base because the tidy version doesn't handle lists as well as base does. i added more comments to explain what's happening tho

bpowers24 commented on Data_Package_Documentation/functions/create_dd.R:
```diff
@@ -0,0 +1,337 @@
+                                      "Unit",  "N/A", "Unit of measurement that applies to a given column or row in the data package.", "text", "column header", '"N/A"; "-9999"; ""; "NA"', "dd template",
+                                      "Definition", "N/A", "Description of the information in a given column or row in the dataset.definition", "text", "column header", '"N/A"; "-9999"; ""; "NA"', "dd template",
+                                      "Data_Type", "N/A", "Type of data (numeric; text; date; time; datetime).", "text", "column header", '"N/A"; "-9999"; ""; "NA"', "dd template", 
+                                      "Term_Type", "N/A", "Indicates whether the header is a column header or row header.", "text", "column header", '"N/A"; "-9999"; ""; "NA"', "dd template", 
```
Confirm this definition is okay.

bpowers24 commented on Data_Package_Documentation/functions/create_dd.R:
```diff
@@ -0,0 +1,337 @@
+      Unit = first(na.omit(Unit)),
+      Definition = first(na.omit(Definition)),
+      Data_Type = first(na.omit(Data_Type)),
+      Term_Type = paste(Term_Type, collapse = ", "),
```
Confirm this is okay. 
e.g., term type for "Unit" = "row header, column header"

bpowers24 commented on Data_Package_Documentation/functions/test_that/test-create_flmd.R:
```diff
@@ -0,0 +1,226 @@
+  add_example_boye(my_data_package_dir)
+  add_example_goldman(my_data_package_dir)
+  
+  my_files <- get_files(directory = my_data_package_dir)
```
Thanks for troubleshooting this!

bpowers24 commented on Data_Package_Documentation/functions/test_that/test-create_flmd.R:
header rows above and below, with and without hashtags, and a mixed one with all of them. add 5 examples

bpowers24 commented:
**A summary of the main changes I made to address your comments:**
## create_flmd_dd.R
- I added comments that give tips for including and excluding files.
## create_dd.R
- I added term type to the DD.
## create_flmd.R
- I refactored the get_files() function to better describe the how the function is including and excluding files.
- I added the methods codes and installation methods to the default standards.
- I changed the function to primarily display the name of the file and then only have it print the df if the user asks.
- I fixed the column or row name position and header rows calculations.
- I changed it so the user indicates a letter rather than a number when specifying the standard to add.
- I rearranged the code to clean up the order and simplify it.
## example_data_for_flmd_dd_tests.R
- I took your suggestion to fix the Windows backslash issue.
- I added a .tsv file to the example data
- I added five more examples of .csv files with header rows with and without hashtags and above or below the column headers.
## test_create_dd.R
- I updated the tests to account for the changes I made; they should pass.
## test_create_flmd.R
- I updated the tests to account for the changes I made; they should pass.
## load_tabular_data.R
- I refactored this function so it works with the get_files input.
- The only required input for this function is files_df. If query_header_info=true, then you can also optionally provide an FLMD. If you don't have header rows, then the FLMD is not needed.
- I was able to run this function on its own and successfully loaded in all of the data. However, I have not checked to see if my changes broke any parts of the data checks workflow. I added a GitHub issue to remind myself to look into this.

brieanne-forbes commented on Data_Package_Documentation/functions/create_flmd.R:
I'm wondering if the flmd/dd/readme exists, if we should add in the definition. Its only added if the placeholder is added rn

brieanne-forbes replied:
I'm okay with this

brieanne-forbes commented on Data_Package_Documentation/functions/create_dd.R:
```diff
@@ -271,22 +269,22 @@ create_dd <- function(files_df,
+                                      "Unit",  "N/A", "Unit of measurement that applies to a given column or row in the data package.", "text", "column header", '"N/A"; "-9999"; ""; "NA"', "dd template",
+                                      "Definition", "N/A", "Description of the information in a given column or row in the dataset.definition", "text", "column header", '"N/A"; "-9999"; ""; "NA"', "dd template",
+                                      "Data_Type", "N/A", "Type of data (numeric; text; date; time; datetime).", "text", "column header", '"N/A"; "-9999"; ""; "NA"', "dd template", 
+                                      "Term_Type", "N/A", "Indicates how the term is used in the data package (e.g., a column header, row header, data flag, or other variable).", '"N/A"; "-9999"; ""; "NA"', "dd template", 
```
```suggestion
                                      "Term_Type", "N/A", "Indicates how the term is used in the data package (e.g., a column header, row header, data flag, or other variable).","text", "column header", '"N/A"; "-9999"; ""; "NA"', "dd template", 
```

brieanne-forbes commented on Data_Package_Documentation/functions/create_dd.R:
something weird is happening with term type
![image](https://github.com/user-attachments/assets/73af6a89-e421-4b2e-ba33-3cf69a7c11fb)

brieanne-forbes commented on Data_Transformation/functions/load_tabular_data.R:
```diff
@@ -152,16 +130,17 @@ load_tabular_data <- function(files_df,
+      files_not_in_dir <- setdiff(flmd_info$File_Path_Absolute, tabular_metadata$all)
       
       if (length(files_not_in_flmd) > 0) {
         log_warn(paste0("The following file is in the directory but NOT in the flmd: ", basename(files_not_in_flmd)))
```
what happens? will the function work for the files not in flmd?

bpowers24 replied:
It's only getting column_or_row_name_position from the FLMD. Anything it can't match to in the FLMD it will ask the user for

brieanne-forbes replied:
I recommend testing how it acts if you have a file thats not in the flmd just to make sure it will function right for that file

bpowers24 replied:
Good catch! This caught a bug where I forgot to name a variable

brieanne-forbes commented on Data_Transformation/functions/load_tabular_data.R:
i provided a flmd but still was asked what line column headers are on and where data starts

brieanne-forbes replied:
were you going to add something here about deal with the error that #end_data causes? or is that for later

brieanne-forbes replied:
Thats what i thought you said and why I commented that it asked me both. I used the test data package I showed you before and an flmd from running the flmd function.

bpowers24 replied:
Nope. this is what brie used: `"Z:\00_ESSDIVE\01_Study_DPs\Test_Data_Package\Test_Data_Package"`

brieanne-forbes commented on Data_Transformation/functions/load_tabular_data.R:
```diff
@@ -237,147 +216,134 @@ load_tabular_data <- function(files_df,
-      # store start data row into the current_df_metadata
-      current_df_metadata <- current_df_metadata %>% 
-        mutate(Data_Start_Row = case_when(.$File_Path_Absolute == current_df_metadata_file_path_absolute ~ current_data_start_row, T ~ Data_Start_Row))
+      log_info("Asking for user input again because prevoius input included an invalid (negative) value. ")
```
```suggestion
      log_info("Asking for user input again because previous input included an invalid (negative) value. ")
```

bpowers24 commented on Data_Package_Documentation/functions/create_flmd.R:
```diff
@@ -414,14 +454,26 @@ create_flmd <- function(files_df, # required
+  ### add standard #############################################################
+  # update the standard based on CSV reporting format keywords (https://github.com/ess-dive-workspace/essdive-file-level-metadata/blob/main/RF_FLMD_Standard_Terms.csv)
+    
+  current_flmd_skeleton <- current_flmd_skeleton %>%
+    mutate(Standard = case_when(tolower(header_format) == "b" ~ "ESS-DIVE Water-Soil-Sediment Chem v1; ESS-DIVE CSV v1", # boye rf
+                                tolower(header_format) == "g" ~ "ESS-DIVE Hydrologic Monitoring v1; ESS-DIVE CSV v1", # goldman rf
+                                str_detect(File_Name, "Methods_Codes\\.csv$") ~ "ESS-DIVE Water-Soil-Sediment Chem v1; ESS-DIVE CSV v1", # boye rf
+                                str_detect(File_Name, "InstallationMethods\\.csv$") ~ "ESS-DIVE Hydrologic Monitoring v1; ESS-DIVE CSV v1", # goldman rf
+                                str_detect(File_Name, "flmd\\.csv$") ~ "ESS-DIVE FLMD v1; ESS-DIVE CSV v1", # flmd rf
+                                str_detect(File_Name, "dd\\.csv$") ~ "ESS-DIVE FLMD v1; ESS-DIVE CSV v1", # flmd rf
+                                str_detect(File_Name, "\\.csv$|\\.tsv$") ~ "ESS-DIVE CSV v1", # csv rf
+                                T ~ "N/A"))
```
what edit did you make here? I thought I had included the addition of methods codes and install methods?

bpowers24 commented:
**A summary of the main changes I made since your last review:**
## create_dd.R
- I fixed the file path to correctly use the parent directory. 
- I accepted your commit suggestions for adding the term type and data type cells that I had missed.
- I fixed the issue where the term type was concatenating NA values incorrectly.
## create_flmd.R
- I accepted your suggestion that fixed the duplication of the parent directory in the file path. 
- I accepted your suggestion to the sample reporting format to the IGSN mapping files.
## load_tabular_data.R
- I addressed the bug you found where the function asks for the line that column headers are on even if you provided an FLMD. The fix should now only ask for the column header position for files that either aren't in the FLMD or are not able to be joined from the FLMD.

brieanne-forbes commented on Data_Package_Documentation/functions/create_flmd.R:
```diff
@@ -473,6 +473,7 @@ create_flmd <- function(files_df, # required
                                 str_detect(File_Name, "dd\\.csv$") ~ "ESS-DIVE FLMD v1; ESS-DIVE CSV v1", # flmd rf
                                 str_detect(File_Name, "\\.csv$|\\.tsv$") ~ "ESS-DIVE CSV v1", # csv rf
+                                str_detect(File_Name, "IGSN_Mapping\\.csv$") ~ "ESS-DIVE Sample v1; ESS-DIVE CSV v1", # sample rf 
                                 T ~ "N/A"))
```
It wont let me add a change to the chunk but IGSN_Mapping needs to be moved above the csv one and InstallationMethods should have an underscores separating

brieanne-forbes commented on Data_Transformation/functions/load_tabular_data.R:
Do you think it makes more sense to ask for the flmd file rather than df? I realized I was making the mistake of giving a path. I'm thinking of using this in the context of data checks - i dont think the flmd will be read in

brieanne-forbes commented on Data_Package_Documentation/functions/test_that/test-create_dd.R:
```diff
@@ -0,0 +1,201 @@
+
+### expected warnings ##########################################################
+
+test_that("expected warnings", {
```
![image](https://github.com/user-attachments/assets/04136a45-0cda-4b60-bd45-c0f73a31b3a0)

brieanne-forbes replied:
Okay to be empty - recommend deleting the test or giving some other output that indicates it passed

brieanne-forbes commented on Data_Package_Documentation/functions/test_that/test-create_dd.R:
all other tests passed

brieanne-forbes commented on Data_Package_Documentation/functions/test_that/test-create_flmd.R:
```diff
@@ -0,0 +1,238 @@
+
+
+### expected typical inputs ####################################################
+test_that("expected typical inputs", {
```
this test failed for me

brieanne-forbes replied:
![image](https://github.com/user-attachments/assets/d4a8c014-70bc-4aac-8814-2d6fc3c1f678)
ran all letter files as unknown

brieanne-forbes commented on Data_Package_Documentation/functions/test_that/test-create_flmd.R:
```diff
@@ -0,0 +1,238 @@
+
+### expected warnings ##########################################################
+
+test_that("expected warnings", {
```
![image](https://github.com/user-attachments/assets/10850fb6-b20b-457f-a94e-50f2b251a584)

brieanne-forbes commented on Data_Package_Documentation/functions/test_that/test-create_flmd.R:
tests passed unless otherwise noted

bpowers24 commented:
**A summary of the changes I made since the last review:**
## create_dd.R
- I added underscores to all of the templated Term_Type values.
- I fixed the error where term type was concatenating NA values incorrectly. It should now return semicolon separated term types if there's more than one or otherwise a single term type or leave the cell empty upon export.
## create_flmd.R
- I fixed the order and typos so it now correctly applies standards to the FLMD template.
- I fixed a bug that was improperly handling the parent folder, especially when the name of folder A was the same as folder B. This required edits to both get_files() and create_flmd() functions. 
## test-create_dd.R
- I wrote a new test to confirm that the Term_Types are concatenating as expected.
- I commented out the expected warnings tests.
## test-create_flmd.R
- I commented out the expected warnings tests.
- I reran the tests and confirmed that they all pass when you enter unknown for all of the lettered files (i.e., file_a-g.csv).
## load_tabular_data.R
- I fixed a bug where the function was asking the user for all header positions even when an FLMD was present. It will now only ask for header positions for files that either aren't in the FLMD or it's not able to join up to.
- I accepted your code suggestions to fix typos.

brieanne-forbes reviewed (approved):
Looks good. Thanks for all of your work on this!
